metric_tank/cluster.go

```diff
@@ -0,0 +1,213 @@
+	return &ClusterStatus{
+		Instance:   instance,
+		Primary:    initialState,
+		LastChange: time.Now(),
+	}
+}
+
+func (c *ClusterStatus) Marshal() ([]byte, error) {
+	c.Lock()
+	defer c.Unlock()
+	return json.Marshal(c)
+}
+
+func (c *ClusterStatus) Set(newState bool) {
+	c.Lock()
+	defer c.Unlock()
+	c.Primary = newState
+	c.LastChange = time.Now()
+}
+
+func (c *ClusterStatus) IsPrimary() bool {
+	c.Lock()
+	defer c.Unlock()
+	return c.Primary
+}
+
+type PersistMessage struct {
+	Instance string `json:"instance"`
+	Key      string `json:"key"`
+	T0       uint32 `json:"t0"`
+}
+
+func (msg *PersistMessage) Send() {
+	if *topicNotifyPersist == "" {
+		return
+	}
+	body, err := json.Marshal(&msg)
+	if err != nil {
+		log.Fatal(4, "failed to marshal persistMessage to json.")
+	}
+	hostPoolResponse := hostPool.Get()
+	p := producers[hostPoolResponse.Host()]
+	err = p.Publish(*topicNotifyPersist, body)
+	if err != nil {
+		log.Warn("publisher marking host %s as faulty due to %s", hostPoolResponse.Host(), err)
+		hostPoolResponse.Mark(err)
+		msg.Send()
+	}
+	return
+}
+
+type MetricPersistHandler struct {
+	metrics Metrics
+}
+
+func NewMetricPersistHandler(metrics Metrics) *MetricPersistHandler {
+	return &MetricPersistHandler{
+		metrics: metrics,
+	}
+}
+
+func (h *MetricPersistHandler) HandleMessage(m *nsq.Message) error {
+	ms := PersistMessage{}
+	err := json.Unmarshal(m.Body, &ms)
+	if err != nil {
+		log.Error(3, "skipping message. %s", err)
+		return nil
+	}
+	if ms.Instance == *instance {
+		log.Debug("skipping message we generated. %s - %s:%d", ms.Instance, ms.Key, ms.T0)
+		return nil
+	}
+
+	// get metric
+	if agg, ok := metrics.Get(ms.Key); ok {
+		// get chunk
+		agg.(*AggMetric).SaveChunkByT0(ms.T0)
+	}
+
+	return nil
+}
+
+func InitCluster(instance string, initialState bool, metrics Metrics, stats met.Backend) {
+	clusterStatus = NewClusterStatus(instance, initialState)
+
+	// init producers
+	pCfg := nsq.NewConfig()
+	pCfg.UserAgent = "metrics_tank"
+	err := app.ParseOpts(pCfg, *producerOpts)
+	if err != nil {
+		log.Fatal(4, "failed to parse nsq producer options. %s", err)
+	}
+	hostPool = hostpool.NewEpsilonGreedy(strings.Split(*nsqdTCPAddrs, ","), 0, &hostpool.LinearEpsilonValueCalculator{})
+	producers = make(map[string]*nsq.Producer)
+
+	for _, addr := range strings.Split(*nsqdTCPAddrs, ",") {
+		producer, err := nsq.NewProducer(addr, pCfg)
+		if err != nil {
+			log.Fatal(4, "failed creating producer %s", err.Error())
+		}
+		producers[addr] = producer
+	}
+
+	// init consumers
+	cfg := nsq.NewConfig()
+	cfg.UserAgent = "metrics_tank_cluster"
+	err = app.ParseOpts(cfg, *consumerOpts)
+	if err != nil {
+		log.Fatal(4, "failed to parse nsq consumer options. %s", err)
+	}
+	cfg.MaxInFlight = *maxInFlight
+
+	consumer, err := insq.NewConsumer(*topicNotifyPersist, fmt.Sprintf("metric_tank_%06d#ephemeral", rand.Int()%999999), cfg, "metric_persist.%s", stats)
```

Comment:
why don't we use a non-ephemeral channel named after the instance? that way in case of disconnect, transient error, .. we can reconnect and get the messages we would otherwise lose.